### PR TITLE
Refactor inband registration (and cancelation)

### PR DESCRIPTION
**Pull request by Dzol**
This began as a task to increase coverage, then became about making Mongoose conform to XEP 0077,  but to do this we need to refactor.

### CODE REVIEW

File: apps/ejabberd/src/mod_register.erl [hunk 1]
```diff
@@ -74,278 +74,288 @@ clean_opt({registration_watchers, Watchers}) ->
...
-		  sub_el = [#xmlel{name = <<"query">>,
-				   attrs = [{<<"xmlns">>, <<"jabber:iq:register">>}],
-				   children = [#xmlel{name = <<"instructions">>,
-				              children = [#xmlcdata{content = TranslatedMsg}]},
-				       #xmlel{name = <<"username">>,
-					      children = UsernameSubels},
-				       #xmlel{name = <<"password">>}
-				       | QuerySubels]}]}
+%% Clients must register before being able to authenticate.
+process_unauthenticated_iq(Sender, Receiver, #iq{type = set} = Stanza, IPAddr) ->
```
Comment (ppikula) [inline apps/ejabberd/src/mod_register.erl]: I'd stick to one standard - either use  From/To or Sender/Receiver, not both alternately. The same story with IQ/Stanza. For me it was slightly misleading when I was comparing `process_unauthenticated_iq` to `process_iq`.

File: apps/ejabberd/src/mod_register.erl [hunk 2]
```diff
@@ -74,278 +74,288 @@ clean_opt({registration_watchers, Watchers}) ->
...
+
+has_only_remove_child(#xmlel{children = C} = Q) when length(C) =:= 1 ->
+    case exml_query:path(Q, [{element, <<"remove">>}], absent) of
+        absent ->
+            {false, absent};
+        _ ->
+            true
+    end;
+has_only_remove_child(#xmlel{children = C} = Q) when length(C) > 1 ->
+    case exml_query:path(Q, [{element, <<"remove">>}], absent) of
```
Comment (ppikula) [inline apps/ejabberd/src/mod_register.erl]: `exml_query:path` takes  leftmost possible described path. I guess the `exml_query:paths` function fits better here, if we want to tell, that the remove element is not present.

File: apps/ejabberd/src/mod_register.erl [hunk 3]
```diff
@@ -74,278 +74,288 @@ clean_opt({registration_watchers, Watchers}) ->
...
+            attempt_cancelation(ClientJID, ServerJID, Stanza);
+        {false, more} ->
+            error_response(Stanza, ?ERR_BAD_REQUEST);
+        {false, absent} ->
+            case has_username_and_password_children(Query) of
+                true ->
+                    Credentials = get_username_and_password_values(Query),
+                    register_or_change_password(Credentials, ClientJID, ServerJID, Stanza, Source);
+                false ->
+                    ignore
```
Comment (ppikula) [inline apps/ejabberd/src/mod_register.erl]: should it be `error_response(..., ?ERR_BAD_REQUEST)`?  The same error as several lines above - L120. Both paths are result of invalid user input, but one of them returns error while the other one just ignores iq.

File: apps/ejabberd/src/mod_register.erl [hunk 4]
```diff
@@ -74,278 +74,288 @@ clean_opt({registration_watchers, Watchers}) ->
...
+has_only_remove_child(#xmlel{children = C} = Q) when length(C) > 1 ->
+    case exml_query:path(Q, [{element, <<"remove">>}], absent) of
+        absent ->
+            {false, absent};
+        _ ->
+            {false, more}
+    end.
+
+has_username_and_password_children(Q) ->
+    {absent, absent} =/= get_username_and_password_elements(Q).
```
Comment (ppikula) [inline apps/ejabberd/src/mod_register.erl]: What if we have just password or just username?  That will give  `{absent, absent} =/= {pawel, absent}` or   `{absent, absent} =/= {absent, admin123}`. Both give true, but according to the function name - has_username_**_and_**_password they shouldn't.

File: apps/ejabberd/src/mod_register.erl [hunk 5]
```diff
@@ -556,4 +566,23 @@ ip_to_integer({IP1, IP2, IP3, IP4}) ->
...
+    true;
+is_query_element(_) ->
+    false.
+
+error_response(Request, Reasons) when is_list(Reasons) ->
+    Request#iq{type = error, sub_el = Reasons};
+error_response(Request, Reason) ->
+    Request#iq{type = error, sub_el = Reason}.
+
+print_it(Text, Value) ->
```
Comment (ppikula) [inline apps/ejabberd/src/mod_register.erl]: Dead code. I guess it is leftover of some debugging

File: test/ejabberd_tests/tests/login_SUITE.erl [hunk 6]
```diff
@@ -173,12 +206,63 @@ register(Config) ->
             escalus:assert_many(Predicates, escalus:wait_for_stanzas(Admin, 2))
         end).
 
+null_password(Config) ->
+    Jimmy = {jimmy,
+             [{username, <<"jimmy">>},
+              {password, <<>>},
+              {server,<<"localhost">>}]},
+    {error, _, Response} = escalus_users:create_user(Config, Jimmy),
+            ct:pal("STANZA = ~p.~n", [Response]),
```
Comment (ppikula) [inline test/ejabberd_tests/tests/login_SUITE.erl]: I'd vote for removing all `ct:pal` calls from this suite

**Comment by michalwski** [comment]
One note from me. Maybe it's good time to move all test related to registration to a separate SUITE?

**Comment by Dzol** [comment]
@michalwski Yeah, sure, I'll make the changes Paweł commented on. Then I'll move them over to their own suite.

File: apps/ejabberd/src/mod_register.erl (hunk 1 above)
Comment (Dzol) [inline apps/ejabberd/src/mod_register.erl]: Okay, changed both: commit 1d811e761557da5ac6f88d56a0ec44dd3aff5635.

File: apps/ejabberd/src/mod_register.erl (hunk 4 above)
Comment (Dzol) [inline apps/ejabberd/src/mod_register.erl]: Fixed in commit 3c2df1027cce7fbe6bd4aed9a19edc777cbf71d6.

File: apps/ejabberd/src/mod_register.erl (hunk 2 above)
Comment (Dzol) [inline apps/ejabberd/src/mod_register.erl]: Fixed in commit b1552d466a465de98f343ffabddfb8c149b5d486.

File: apps/ejabberd/src/mod_register.erl (hunk 3 above)
Comment (Dzol) [inline apps/ejabberd/src/mod_register.erl]: Commit 067168ae9ba90c78262b191d0dcf7a3947573820 makes this more consistent.

File: apps/ejabberd/src/mod_register.erl [hunk 7]
```diff
@@ -74,278 +74,288 @@ clean_opt({registration_watchers, Watchers}) ->
...
+    case inband_registration_and_cancelation_allowed(ServerDomain, ClientJID) of
+        true ->
+            #iq{sub_el = Children, lang = Lang} = IQ,
+            try_register_or_set_password(Username, ServerDomain, Password, ClientJID, IQ, Children, IPAddr, Lang);
+        false ->
+            %% This is not described in XEP 0077.
+            error_response(IQ, ?ERR_FORBIDDEN)
+    end.
+
+attempt_cancelation(ClientJID, #jid{lserver = ServerDomain}, #iq{id = ID, sub_el = Child} = IQ) ->
```
Comment (lavrin) [inline apps/ejabberd/src/mod_register.erl]: cancelation -> cancellation

Comment (lavrin) [inline apps/ejabberd/src/mod_register.erl]: Oh, sorry, the dictionary adds that _US also cancelation_ :) My mistake.

File: apps/ejabberd/src/mod_register.erl [hunk 8]
```diff
@@ -74,278 +74,288 @@ clean_opt({registration_watchers, Watchers}) ->
...
-				       #xmlel{name = <<"password">>}
-				       | QuerySubels]}]}
+%% Clients must register before being able to authenticate.
+process_unauthenticated_iq(From, To, #iq{type = set} = IQ, IPAddr) ->
+    process_iq_set(From, To, IQ, IPAddr);
+process_unauthenticated_iq(From, To, #iq{type = get} = IQ, IPAddr) ->
+    process_iq_get(From, To, IQ, IPAddr).
+
+process_iq(From, To, #iq{type = set} = IQ) ->
+    process_iq_set(From, To, IQ, jlib:jid_tolower(From));
```
Comment (michalwski) [inline apps/ejabberd/src/mod_register.erl]: Please take a look at `jid` module and change all the deprecated functions from `jlib`.

**Comment by michalwski** [comment]
How about [`try_set_password`](https://coveralls.io/builds/4241734/source?filename=apps%2Fejabberd%2Fsrc%2Fmod_register.erl#L226) function. I think a test case for that should be easy.

**Comment by michalwski** [comment]
Another test would be for the [`welcome message`](https://coveralls.io/builds/4241734/source?filename=apps%2Fejabberd%2Fsrc%2Fmod_register.erl#L311). This will require some `mod_register` config changes in `init_per_testcase` and reverting these changes in `end_per_testcase`.

**Comment by Dzol** [comment]
> How about try_set_password function. I think a test case for that should be easy.

Commit 2ba5082f69cf437c1a64eea7c98382790c704b9c tests this, the subsequent commit 0f6878a450ea9dac8b5ec70f558ad8bd7a1e4786 fixes it.

**Comment by Dzol** [comment]
> Another test would be for the welcome message. This will require some mod_register config changes in init_per_testcase and reverting these changes in end_per_testcase.

@michalwski Do you think testing registration timeout (that looks like inter-registration time to me) and some of the IP access control would do more for coverage?

File: apps/ejabberd/src/ejabberd_auth.erl [hunk 9]
```diff
@@ -227,6 +227,8 @@ check_digest(Digest, DigestGen, _Password, Passwd) ->
                    Server :: ejabberd:server(),
                    Password :: binary()
                   ) -> ok | {error, empty_password | not_allowed | invalid_jid}.
+set_password(_,_, <<"">>) ->
+    {error, empty_password};
 set_password(_User, _Server, "") ->
```
Comment (michalwski) [inline apps/ejabberd/src/ejabberd_auth.erl]: I'd remove the head with empty string `""` as it's never called (and shouldn't be).

Comment (Dzol) [inline apps/ejabberd/src/ejabberd_auth.erl]: Commit 17fae79e2afe211aeab8a19e0a86bc6ab75dc8fc fixes this.

File: apps/ejabberd/src/mod_register.erl (hunk 5 above)
Comment (Dzol) [inline apps/ejabberd/src/mod_register.erl]: Removed it.

File: test/ejabberd_tests/tests/login_SUITE.erl (hunk 6 above)
Comment (Dzol) [inline test/ejabberd_tests/tests/login_SUITE.erl]: Removed these too.

File: apps/ejabberd/src/mod_register.erl (hunk 8 above)
Comment (Dzol) [inline apps/ejabberd/src/mod_register.erl]: Commit 2f1765b8d49cbfdd3ad5c6d3ee1fd9ea6db27ca2 fixes these.

**Pull request merged**